File: rest-service/src/main/java/mb/lib/model/JobStatus.java
package mb.lib.model;

import java.util.Optional;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

public enum JobStatus
{
  Queued("queued"),
  InProgress("in-progress"),
  Errored("errored"),
  Completed("completed"),
  Expired("expired");

  private static final String InProgressAlias = "grabbed";
  private static final String QueuedAlias     = "claimed";

  private final String value;

  JobStatus(String name) {
    this.value = name;
  }

  @JsonValue
  public String getValue() {
    return value;
  }

  public static Optional<JobStatus> fromString(String name) {
    for (var e : values())
      if (e.value.equals(name))
        return Optional.of(e);

    return Optional.ofNullable(switch (name) {
      case QueuedAlias -> Queued;
      case InProgressAlias -> InProgress;
      default -> null;
    });
  }

  @JsonCreator
  public static JobStatus unsafeFromString(String name) {
    return fromString(name)
      .orElseThrow(() -> new IllegalArgumentException("Unrecognized JobStatus value: " + name));
  }
}

File: rest-service/src/main/java/mb/lib/model/JobStatus.kt
package mb.lib.model

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.TextNode
import java.util.*

enum class JobStatus(val value: String) {
  Queued     ("queued"),
  InProgress ("in-progress"),
  Errored    ("errored"),
  Completed  ("completed"),
  Expired    ("expired");

  @JsonValue
  fun toJSON(): JsonNode = TextNode(value)

  companion object {
    private const val InProgressAlias = "grabbed"
    private const val QueuedAlias     = "claimed"

    fun fromString(name: String): Optional<JobStatus> {
      for (e in values())
        if (e.value == name)
          return Optional.of(e)

      return Optional.ofNullable(when (name) {
        QueuedAlias     -> Queued
        InProgressAlias -> InProgress
        else            -> null
      })
    }

    @JsonCreator
    fun unsafeFromString(name: String): JobStatus =
      fromString(name).orElseThrow { IllegalArgumentException("Unrecognized JobStatus value: $name") }
  }
}
